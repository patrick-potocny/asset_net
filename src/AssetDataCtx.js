import { createContext } from "react";

export const AssetDataCtx = createContext()