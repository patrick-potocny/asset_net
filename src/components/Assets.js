import React, { useContext, useEffect, useState } from "react";
import { v4 as uuid } from "uuid";
import PropTypes from "prop-types";
import AssetCard from "./AssetCard";
import Modal from "./Modal";
import AddAsset from "./AddAsset";
import plusIcon from "../assets/images/plus.svg";
import { AssetDataCtx } from "../AssetDataCtx";

function Assets({ assetType }) {
  const { assetData } = useContext(AssetDataCtx);
  const [isOpen, setIsOpen] = useState(false);
  const [displayedAssets, setDisplayedAssets] = useState([]);
  
  useEffect(() => {
    setDisplayedAssets(
      assetData.filter((asset) => asset.assetType === assetType)
    );
    if (assetType === "all_assets") setDisplayedAssets(assetData);
  }, [assetData, assetType]);

  return (
    <div className="assets">
      {displayedAssets.map((asset) => (
        <AssetCard key={uuid()} asset={asset} />
      ))}

      <div
        className="assets__add"
        onClick={() => {
          setIsOpen(true);
        }}
      >
        <img src={plusIcon} alt="Add" />
      </div>

      <Modal isOpen={isOpen} onClose={() => setIsOpen(false)}>
        <AddAsset onClose={() => setIsOpen(false)} />
      </Modal>
    </div>
  );
}

Assets.propTypes = {
  assetType: PropTypes.string.isRequired,
};

export default Assets;
