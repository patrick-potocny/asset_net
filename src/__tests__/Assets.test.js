/* eslint-disable */
import React from "react";
import { render, screen } from "@testing-library/react";
import { AssetDataCtx } from "../AssetDataCtx";
import Assets from "../components/Assets";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";

jest.mock("../components/AssetCard", () => ({ asset }) => (
  <div data-testid="asset-card">{asset.name}</div>
));

jest.mock("../components/Modal", () => ({ isOpen, onClose }) => (
  <div data-testid="mock-modal">isOpen is {isOpen.toString()}</div>
));

const mockAssetData = [
  { name: "BTC", assetType: "crypto" },
  { name: "AAPL", assetType: "stocks" },
  { name: "MSFT", assetType: "stocks" },
];

describe("Assets component", () => {
  test('should render properly', () => {
    const { container } = render(
      <AssetDataCtx.Provider value={{ assetData: mockAssetData }}>
        <Assets assetType="stocks" />
      </AssetDataCtx.Provider>
    );

    expect(container).toMatchSnapshot()
  });

  test("should render assets based on assetType", () => {
    render(
      <AssetDataCtx.Provider value={{ assetData: mockAssetData }}>
        <Assets assetType="stocks" />
      </AssetDataCtx.Provider>
    );

    expect(screen.queryAllByTestId("asset-card").length).toBe(2);
    expect(screen.getByText("AAPL")).toBeInTheDocument();
  });

  test("should display all assets", () => {
    render(
      <AssetDataCtx.Provider value={{ assetData: mockAssetData }}>
        <Assets assetType="all_assets" />
      </AssetDataCtx.Provider>
    );

    expect(screen.queryAllByTestId("asset-card").length).toBe(3);
    expect(screen.getByText("BTC")).toBeInTheDocument();
    expect(screen.getByText("AAPL")).toBeInTheDocument();
  });

  test("should change modals prop after clicking on add button", () => {
    render(
      <AssetDataCtx.Provider value={{ assetData: mockAssetData }}>
        <Assets assetType="stocks" />
      </AssetDataCtx.Provider>
    );

    expect(screen.queryByText('isOpen is false')).toBeInTheDocument()

    userEvent.click(screen.getByAltText('Add'))

    expect(screen.queryByText('isOpen is true')).toBeInTheDocument()
  });
});
