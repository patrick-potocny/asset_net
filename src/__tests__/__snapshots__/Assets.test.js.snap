// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assets component should render properly 1`] = `
<div>
  <div
    class="assets"
  >
    <div
      data-testid="asset-card"
    >
      AAPL
    </div>
    <div
      data-testid="asset-card"
    >
      MSFT
    </div>
    <div
      class="assets__add"
    >
      <img
        alt="Add"
        src="plus.svg"
      />
    </div>
    <div
      data-testid="mock-modal"
    >
      isOpen is 
      false
    </div>
  </div>
</div>
`;
